From a6b53988e92eb128ba022a050e18f16c5644afb2 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:01:11 -0400
Subject: [fs] btrfs: skip readonly root for snapshot-aware defragment

Message-id: <1407179109-13714-185-git-send-email-sandeen@redhat.com>
Patchwork-id: 87302
O-Subject: [RHEL7.1 PATCH 184/422] Btrfs: skip readonly root for snapshot-aware defragment
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Wang Shilong <wangsl.fnst@cn.fujitsu.com>

commit bcbba5e6593281adc234938b42d3c3d3570335db
Author: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
Date:   Sat Feb 8 23:46:35 2014 +0800

    Btrfs: skip readonly root for snapshot-aware defragment

    Btrfs send is assuming readonly root won't change, let's skip readonly root.

    Signed-off-by: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index e534833..3441329 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -2240,6 +2240,11 @@ static noinline int relink_extent_backref(struct btrfs_path *path,
   return PTR_ERR(root);
  }
 
+ if (btrfs_root_readonly(root)) {
+  srcu_read_unlock(&fs_info->subvol_srcu, index);
+  return 0;
+ }
+
  /* step 2: get inode */
  key.objectid = backref->inum;
  key.type = BTRFS_INODE_ITEM_KEY;
-- 
1.7.1